### PR TITLE
Work with Cloud Storage OCN(Object Change Notification)

### PR DESCRIPTION
**Pull request by akm**
Use [OCN](https://cloud.google.com/storage/docs/object-change-notification) instead of self difference checking.

## Changes

- Accept `GET` and `POST` requests to root path
    - `GET` is called in order to check the app has valid site verification.
    - `POST is called to notify changes
- Don't use `datastore`, but just publish the notified change to the pubsub topic.
- Add descriptions to setup to [README.md](https://github.com/groovenauts/blocks-gcs-watcher/blob/45415c6aebdd4a5967400b94d1a64bd5b3f7f87b/README.md)

### CODE REVIEW

**Review by hokuts — APPROVED** [review]

File: pubsub_notifier.go [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+		Messages: []*pubsub.PubsubMessage{msg},
+	}
+	return pp.topicsService.Publish(topic, req).Do()
+}
+
+func NewPubsubNotifier(ctx context.Context) (Notifier, error) {
+	// https://github.com/google/google-api-go-client#application-default-credentials-example
+	client, err := google.DefaultClient(ctx, pubsub.PubsubScope)
+	if err != nil {
+		log.Errorf(ctx, "Failed to create DrfaultClient\n")
```
Comment (minimum2scp) [inline pubsub_notifier.go]: s/DrfaultClient/DefaultClient/ ?

File: processor.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+	var obj map[string]interface{}
+	err = json.Unmarshal(bytes, &obj)
+	if err != nil {
+		return err
+	}
+	log.Infof(ctx, "%v\n", obj)
+
+	bucket := obj["bucket"].(string)
+	name := obj["name"].(string)
```
Comment (minimum2scp) [inline processor.go]: In general, type assertion without multiple-value may trigger panic. Is it ok?

File: pubsub_notifier.go (hunk 1 above)
Comment (akm) [reply pubsub_notifier.go]: @minimum2scp Yes, I fixed it. Thanks.

File: processor.go (hunk 2 above)
Comment (akm) [reply processor.go]: @minimum2scp Thanks a lot for your suggestion. I added the tests and fixed them.

**Review by minimum2scp — APPROVED** [review]

**Review by taigou — APPROVED** [review]

File: handler.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	return c.HTML(http.StatusOK, res)
+}
+
+func (h *handler) post(c echo.Context) error {
+	req := c.Request()
+	ctx := appengine.NewContext(req)
+	log.Infof(ctx, "Processing OCN POST request\nHeader: %v\n", req.Header)
+	resource_state := req.Header.Get("X-Goog-Resource-State")
+	if resource_state == "" {
+		log.Infof(ctx, "Unkown message received.\n")
```
Comment (ko1nksm) [inline handler.go]: Unknown ?

Comment (akm) [reply handler.go]: @ko1nksm Yes. If the API is modified, `X-Goog-Resource-State` header might not be used. So I added a check whether the header exists.
And this check is useful for developer to remember to add the header. Does this answer make a sense?

Comment (akm) [reply handler.go]: @ko1nksm It was a typo. I fixed it. Thanks.

**Review by ko1nksm — APPROVED** [review]

**Review by totty — APPROVED** [review]

**Review by nagachika — APPROVED** [review]

**Review by guemon — APPROVED** [review]

**Pull request merged**